docs/proposals/formatted-output.md

```diff
@@ -0,0 +1,325 @@
+
+- Use shell variables on Unix
+
+```bash
+REFERENCE_A=$(oras push $REGISTRY/$REPO:$TAG hello.txt --format '{{.Ref}}')
+REFERENCE_B=$(oras attach --artifact-type sbom/example $REFERENCE_A sbom.spdx --format '{{.Ref}}') 
+notation sign $REFERENCE_B
+```
+
+- Use [ConvertFrom-Json](https://learn.microsoft.com/powershell/module/microsoft.powershell.utility/convertfrom-json) on Windows PowerShell
+
+```powershell
+$A=oras push $REGISTRY/$REPO:$TAG hello.txt --format json --no-tty | ConvertFrom-Json
+$B=oras attach --artifact-type sbom/example $A.Ref sbom.spdx --format json --no-tty | ConvertFrom-Json
+notation sign $B.Ref
+```
+
+### CI/CD
+
+Bob is a DevOps engineer. He uses the ORAS GitHub Actions [Setup action](https://github.com/oras-project/setup-oras) to install ORAS in his CI/CD workflow. He wants to chain several ORAS commands in a Shell script to perform multiple operations.
+
+For example, pull multiple files (layers) from a repository and filter out the file path of its first layer in the standard output. Then, pass the pulled first layer to the second command (`docker import`) to perform further operations. 
+
+```yaml
+jobs:
+  example-job:
+    steps:
+      - uses: oras-project/setup-oras@v1
+      - run: |
+          PATH=`oras pull $REGISTRY/$REPO:$TAG --format '{{(first .Files).Path}}'`
+          docker import $PATH
+```
+
+## Proposal and desired user experience
+
+Enable users to use the `--format <type>` flag to format output into structured data (e.g. JSON) and use the `--format` with the [Go template](https://pkg.go.dev/text/template) language to manipulate the output fields. Users can still use `--format '{{toRawJson .}}'` to get raw JSON output.
+
+### oras pull 
+
+When using `oras pull` with the flag `--format`, the following fields should be formatted into JSON output:
+
+- `Ref`: full artifact reference by digest, e.g, `$REGISTRY/$REPO@$DIGEST`
+-  `Files`: a list of downloaded files
+    - `Path`: the absolute file path of the pulled file (layer)
+    - `Ref`: full reference by digest of the pulled file (layer)
+    - `MediaType`: media type of the pulled file (layer) 
+    - `Digest`: digest of the pulled file (layer) 
+    - `Size`: file size in bytes
+    - `Annotations`: contains arbitrary metadata for the image manifest
+
+For example, pull an artifact that contains multiple layers (files) and show their descriptor metadata as pretty JSON in standard output:
+
+```bash
+oras pull $REGISTRY/$REPO:$TAG --artifact-type example/sbom sbom.spdx  --artifact-type example/vul-scan vul-scan.json --format json
+```
+
+```json
+{
+  "Ref": "localhost:5000/oras@sha256:7414904f07f515f48fe4afeaf876e3151039a81e7177b9c66e9e7ed6dd186111",
+  "Files": [
```

Comment:
Would recommend all these properties to be lowercase aligns with things like 'mediaType' and other properties in the manifest that are all lower case. @FeynmanZhou

Reply:
@sajayantony Due to the lower casing field selectors are not supported natively via Go template, when use the flag `--format` to print the prettified JSON format, the first letter of selectors and all output fields are supposed to be upper case except for `oras manifest fetch`. 

`oras manifest fetch` formatted output is lower case since its field names are lower case in the OCI Spec and the original output is JSON format.

Reply:
Given our discussion of kubectl, I'm more inclined towards a lower case only and avoid the extra cognitive steps. 
For e.g. why should a user have to thinks about `Size` in the CLI output and `size` in the manifest

Reply:
Agree with using lowercase for the first letter of each property as we chose to follow [kubectl](https://kubernetes.io/docs/reference/kubectl/quick-reference/#viewing-and-finding-resources)'s design principle. Now all properties in the samples have been updated to lowercase.